### DISCUSSION TITLE
Make Pyo3 exposed class pydantic compatible

### DISCUSSION THREAD

**Discussion by GasPasChaud** [pydantic/pydantic#8854] [Question]
Hello there 🖖🏼 

I'm working on a mixed rust/python project using PyO3.

I have a struct/class defined (let's call it `MyResults`) in rust that is representing a model and is the return type of one my rust algorithm. It is used in python for further post-processing. 

In python I'm relying on Pydantic V2 to manipulate my models.

If I try to use the `Results` as is as a value in one of my Pydantic models, I get the error the following error:
```
raise PydanticSchemaGenerationError(
pydantic.errors.PydanticSchemaGenerationError: 
Unable to generate pydantic-core schema for <class 'builtins.MyResults'>. 
Set `arbitrary_types_allowed=True` in the model_config to ignore this error or 
implement `__get_pydantic_core_schema__` on your type to fully support it.

If you got this error by calling handler(<some type>) within `__get_pydantic_core_schema__` 
then you likely need to call `handler.generate_schema(<some type>)` since we do not call 
`__get_pydantic_core_schema__` on `<some type>` otherwise to avoid infinite recursion.
```

I first tried the `arbitrary_types_allowed` flag, but as soon as I tried to serialize my model, it failed.

So here comes the `Annotated` version of `MyResults`

```py
from .my_results import MyResults
from pydantic import BaseModel, PlainSerializer, PlainValidator
from typing_extensions import Annotated


def validate_results(v, _) -> MyResults:
    if isinstance(v, MyResults):
        return v
    if isinstance(v, dict):
        return MyResults(**v)

    raise TypeError("MyResults or dict required")


def serialize_results(
    v: MyResults,
) -> ResultsDict:
    return  v.dict() # implemented in rust


Results = Annotated[
    MyResults,
    PlainValidator(validate_results),
    PlainSerializer(serialize_results),
]
```

Great! Now I can use `Results` everywhere I used `MyResults` in Pydantic models.

However, my colleagues now have to import `Results` and `MyResults` which is a bit cumbersome...

I would like to make `MyResults` pydantic compliant natively in rust.

In Python it would be implementing `__get_pydantic_core_schema__`. But looking at the expected signature, I don't know how to implement it in rust.

As anyone figured this out? Is it possible at all? 

Should I instead define the `MyResults` class in Python as Pydantic model and try to return it from rust functions?

EDIT: formatting of error message for readability

**Comment by Distortedlogic** [comment]
bump

**Comment by LockedThread** [comment]
bump

**Comment by keisanng** [comment]
bump

**Comment by Laurence-Cullen** [comment]
bump!

**Comment by ppmpreetham** [comment]
bump!

**Comment by Lemminkyinen** [comment]
bump!
